src/uct/cuda/cuda_ipc/cuda_ipc_ep.c

```diff
@@ -38,10 +44,55 @@ UCS_CLASS_DEFINE_DELETE_FUNC(uct_cuda_ipc_ep_t, uct_ep_t);
 #define uct_cuda_ipc_trace_data(_addr, _rkey, _fmt, ...)     \
     ucs_trace_data(_fmt " to %"PRIx64"(%+ld)", ## __VA_ARGS__, (_addr), (_rkey))
 
+void *uct_cuda_ipc_ep_attach_rem_seg(uct_cuda_ipc_ep_t *ep,
+                                     uct_cuda_ipc_iface_t *iface,
+                                     uct_cuda_ipc_key_t *rkey)
+{
+    unsigned int cuda_ipc_mh_flags = CU_IPC_MEM_LAZY_ENABLE_PEER_ACCESS;
+    uct_cuda_ipc_rem_seg_t *rem_seg, search;
+    ucs_status_t status;
+
+    search.ph = rkey->ph;
+    /* TODO: Address the case when va matches but not memhandle
+
+       Cause:
+       cudaMalloc(&ptrX ...); mhX = getmemhandle(ptrX, ...); cudafree(ptrX);
+       cudaMalloc(&ptrY ...); mhY = getmemhandle(ptrX, ...);
+       Now: ptrX can be ptrY but mhX is never mhY
+    */
+    rem_seg =
+        sglib_hashed_uct_cuda_ipc_rem_seg_t_find_member(ep->rem_segments_hash,
+                                                        &search);
+    if (rem_seg == NULL) {
+        rem_seg = ucs_malloc(sizeof(*rem_seg), "rem_seg");
+        if (rem_seg == NULL) {
+            ucs_error("failed to allocate memory for a remote segment. %m");
+            return NULL;
+        }
+
+        rem_seg->ph      = rkey->ph;
+        rem_seg->dev_num = rkey->dev_num;
+
+        status =
+            UCT_CUDADRV_FUNC(cuIpcOpenMemHandle((CUdeviceptr *)&rem_seg->d_bptr,
+                                                rkey->ph, cuda_ipc_mh_flags));
+        if (UCS_OK != status) {
+            return NULL;
```

Comment:
leak

Reply:
@yosefe I've addressed the leak and I've added sglib init/teardown now (I had missed this and I'm glad you pointed to this leak).